File: Answer/3606.优惠券校验器.java
/*
 * @lc app=leetcode.cn id=3606 lang=java
 *
 * [3606] 优惠券校验器
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Pattern;

class Solution {
    public List<String> validateCoupons(String[] code, String[] businessLine, boolean[] isActive) {
        HashMap<String, List<String>> res = new HashMap<>();
        res.put("electronics", new ArrayList<>());
        res.put("grocery", new ArrayList<>());
        res.put("pharmacy", new ArrayList<>());
        res.put("restaurant", new ArrayList<>());
        
        int n = code.length;
        
        Pattern pattern = Pattern.compile("^[a-zA-Z0-9_]+$");
        
        for (int i = 0; i < n; i++) {
            if (!isActive[i]) {
                continue;
            }
            if (!res.containsKey(businessLine[i])) {
                continue;
            }
            if (!pattern.matcher(code[i]).matches()) {
                continue;
            }
            res.get(businessLine[i]).add(code[i]);
        }
        
        List<String> ans = new ArrayList<>();
        Collections.sort(res.get("electronics"));
        ans.addAll(res.get("electronics"));
        Collections.sort(res.get("grocery"));
        ans.addAll(res.get("grocery"));
        Collections.sort(res.get("pharmacy"));
        ans.addAll(res.get("pharmacy"));
        Collections.sort(res.get("restaurant"));
        ans.addAll(res.get("restaurant"));
        
        return ans;
    }
}
// @lc code=end

File: Answer/3606.优惠券校验器.py
#
# @lc app=leetcode.cn id=3606 lang=python3
#
# [3606] 优惠券校验器
#

# @lc code=start
class Solution:
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        res = { "electronics": [], "grocery": [], "pharmacy": [], "restaurant": [] }
        n = len(code)
        for i in range(n):
            if not isActive[i]:
                continue
            if businessLine[i] not in res.keys():
                continue
            if not match("^[a-zA-Z0-9_]+$", code[i]):
                continue
            res[businessLine[i]].append(code[i])
        ans = []
        for _, v in res.items():
            v.sort()
            for i in v:
                ans.append(i)
        return ans
        
# @lc code=end

